### PR TITLE
`Compute` - Upgrade API version from `2021-11-01` to `2024-03-01`

### PR DESCRIPTION
**Pull request by ms-zhenhua**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

Upgrade Compute to the latest API version to support new features

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [ ] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

**Note: 3 testcases also fail in main branch. Other failed testcases should be occasional and have passed in local re-run listed below.**

image

--- PASS: TestAccManagedDisk_storageAccountType (213.38s)
--- PASS: TestAccLinuxVirtualMachine_orchestratedZonal (393.79s)
--- PASS: TestAccWindowsVirtualMachineScaleSet_scalingAutoScale (432.61s)
--- PASS: TestAccDiskEncryptionSet_disablePurgeProtection (462.94s)
--- PASS: TestAccLinuxVirtualMachine_imageFromCommunitySharedImageGallery (1692.02s)
PASS

--- PASS: TestAccLinuxVirtualMachineScaleSet_disksDataDiskStorageAccountTypePremiumV2LRS (362.88s)
--- PASS: TestAccOrchestratedVirtualMachineScaleSet_disksDataDiskStorageAccountTypePremiumV2LRS (377.52s)
--- PASS: TestAccWindowsVirtualMachineScaleSet_disksDataDiskStorageAccountTypePremiumV2LRS (410.91s)
--- PASS: TestAccLinuxVirtualMachineScaleSet_disksDataDiskStorageAccountTypePremiumV2LRSWithIOPSAndMBPS (532.81s)
--- PASS: TestAccWindowsVirtualMachineScaleSet_disksDataDiskStorageAccountTypePremiumV2LRSWithIOPSAndMBPS (606.45s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/compute       618.869s

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `Compute` - Upgrade API version from `2021-11-01` to `2024-03-01`

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/compute/client/client.go [hunk 1]
```diff
@@ -15,7 +15,6 @@ import (
 	"github.com/hashicorp/go-azure-sdk/resource-manager/compute/2021-11-01/dedicatedhostgroups"
 	"github.com/hashicorp/go-azure-sdk/resource-manager/compute/2021-11-01/dedicatedhosts"
 	"github.com/hashicorp/go-azure-sdk/resource-manager/compute/2021-11-01/sshpublickeys"
```
Comment (tombuildsstuff) [inline internal/services/compute/client/client.go:17]: Can we also update the other `2021-11-01` resources to use this updated API Version at the same time?

Comment (ms-zhenhua) [reply internal/services/compute/client/client.go]: Hi @tombuildsstuff, I have upgraded other `2021-11-01` resources and the test coverage of `compute` is the same as the `main` branch. Could you kindly have another review? Thanks.

**Comment by tombuildsstuff** [comment]
Dependent on https://github.com/hashicorp/terraform-provider-azurerm/pull/25416

**Review by stephybun — COMMENTED** [review]
@ms-zhenhua could you please rebase this PR to pull in the changes from https://github.com/hashicorp/terraform-provider-azurerm/pull/25416

**Comment by ms-zhenhua** [comment]
Hi @stephybun, I have resolved the conflicts. Could you kindly help take another review?

**Review by stephybun — APPROVED** [review]
Thanks @ms-zhenhua, tests look good. LGTM 👍

**Pull request merged**